### PR TITLE
Decouple the building of benchmarks and tests

### PR DESCRIPTION
**Pull request by leoluan2009**
Tests and benchmarks targets are now de-coupled.
That means they can be built independently. 
Shared functionality is moved to a common utility library.

Resolves https://github.com/facebookincubator/velox/issues/1704

### CODE REVIEW

**Comment by leoluan2009** [comment]
cc @mbasmanova @kgpai  @majetideepak can  you help to review this pr?  the build error is due to https://github.com/facebookincubator/velox/pull/2442

**Comment by kgpai** [comment]
Hi @leoluan2009 , Can you give a brief description of how you have decoupled them ? Thanks !

**Comment by leoluan2009** [comment]
> Hi @leoluan2009 , Can you give a brief description of how you have decoupled them ? Thanks !
use VELOX_ENABLE_BENCHMARKS to control whether to build benchmarks directory and use VELOX_BUILD_TESTING to control whether to build tests directory, if some benchmarks use import the files from tests directory, Add the files to utils subdirectory in tests.  when VELOX_BUILD_TESTING is on or VELOX_BUILD_TEST_UTILS is on, will build tests/utils.

**Comment by kgpai** [comment]
Thanks again @leoluan2009 , Based on my perusal, I see the following changes:
 * Move common test utilities to a separate library.
 * Create separate flag for Benchmarks and decouple benchmark building from tests.
 
Some questions:
* Are benchmarks still built along with tests ? If not how should they be invoked ?
* We run benchmark comparisons to catch performance regressions on merge to main ; We need to ensure that these dont break , due to these changes.

**Comment by leoluan2009** [comment]
> Are benchmarks still built along with tests ? If not how should they be invoked ?

not along with tests, use VELOX_ENABLE_BENCHMARKS or VELOX_ENABLE_BENCHMARKS_BASIC to control whether to build benchmarks, and if one of this two flags is on, set VELOX_BUILD_TEST_UTILS to on for import some common test utilities

**Comment by leoluan2009** [comment]
> We run benchmark comparisons to catch performance regressions on merge to main ; We need to ensure that these dont break , due to these changes. 

This is still controlled by VELOX_ENABLE_BENCHMARKS_BASIC

**Comment by kgpai** [comment]
Thanks @leoluan2009 , Can you make changes here : https://github.com/facebookincubator/velox/blob/main/.circleci/config.yml#L247 so that the benchmarks builds are not broken in this PR.

**Comment by majetideepak** [comment]
@leoluan2009 thank you for this work. I actually have an issue open here https://github.com/facebookincubator/velox/issues/1704

**Comment by leoluan2009** [comment]
> @leoluan2009 thank you for this work. I actually have an issue open here https://github.com/facebookincubator/velox/issues/1704

yes，I think I already follow the design for share files in https://github.com/facebookincubator/velox/issues/1704

**Review by majetideepak — APPROVED** [review]

**Comment by majetideepak** [comment]
@kgpai the benchmarks are still running with this change.
The benchmark target uses the right option.
https://github.com/facebookincubator/velox/blob/d9f13737b0f90df32d52e687d2287f0f3b9c4836/Makefile#L95-L96

https://app.circleci.com/pipelines/github/facebookincubator/velox/12821/workflows/44ace784-d480-4dfe-9d36-fbc6d928d8da/jobs/76628

**Comment by majetideepak** [comment]
@leoluan2009  nit: I guess we can remove the TODO here https://github.com/facebookincubator/velox/blob/main/CMakeLists.txt#L326

**Comment by leoluan2009** [comment]
> @leoluan2009 nit: I guess we can remove the TODO here https://github.com/facebookincubator/velox/blob/main/CMakeLists.txt#L326

Done

**Comment by majetideepak** [comment]
@kgpai @Yuhta can you please merge this? Thanks.

**Comment by leoluan2009** [comment]
cc @majetideepak @kgpai @Yuhta

File: velox/functions/prestosql/tests/utils/FunctionBaseTest.cpp [hunk 1]
```diff
@@ -13,7 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-#include "velox/functions/prestosql/tests/FunctionBaseTest.h"
+#include "FunctionBaseTest.h"
```
Comment (majetideepak) [inline velox/functions/prestosql/tests/utils/FunctionBaseTest.cpp:16]: @kgpai Can you fix this internally now that you imported it? I missed this thanks.

File: velox/vector/tests/utils/VectorMaker.cpp [hunk 2]
```diff
@@ -13,7 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-#include "velox/vector/tests/VectorMaker.h"
+#include "VectorMaker.h"
```
Comment (majetideepak) [inline velox/vector/tests/utils/VectorMaker.cpp:16]: `velox/vector/tests/utils/VectorMaker.h`

File: velox/vector/tests/utils/VectorMaker.h [hunk 3]
```diff
@@ -720,4 +720,4 @@ class VectorMaker {
 
 } // namespace facebook::velox::test
 
-#include "velox/vector/tests/VectorMaker-inl.h"
+#include "VectorMaker-inl.h"
```
Comment (majetideepak) [inline velox/vector/tests/utils/VectorMaker.h:723]: `velox/vector/tests/utils/VectorMaker-inl.h`

File: velox/functions/prestosql/tests/utils/FunctionBaseTest.cpp (hunk 1 above)
Comment (kgpai) [reply velox/functions/prestosql/tests/utils/FunctionBaseTest.cpp:16]: Sure, How does it build then if this is wrong?

File: velox/vector/tests/utils/VectorTestBase.cpp [hunk 4]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-#include "velox/vector/tests/VectorTestBase.h"
+#include "VectorTestBase.h"
```
Comment (majetideepak) [inline velox/vector/tests/utils/VectorTestBase.cpp:17]: `velox/vector/tests/utils/VectorTestBase.h`

**Comment by kgpai** [comment]
@majetideepak I will change bunch of places where relative path is used instead of absolute path.

**Comment by majetideepak** [comment]
> @majetideepak I will change bunch of places where relative path is used instead of absolute path.

@kgpai thank you.

File: velox/functions/prestosql/tests/utils/FunctionBaseTest.cpp (hunk 1 above)
Comment (leoluan2009) [reply velox/functions/prestosql/tests/utils/FunctionBaseTest.cpp]: Done, @majetideepak @kgpai , thanks!

File: velox/vector/tests/utils/VectorMaker.h (hunk 3 above)
Comment (leoluan2009) [reply velox/vector/tests/utils/VectorMaker.h]: Done, @majetideepak , thanks!

File: velox/vector/tests/utils/VectorTestBase.cpp (hunk 4 above)
Comment (leoluan2009) [reply velox/vector/tests/utils/VectorTestBase.cpp:17]: Done, Thanks!

File: velox/vector/tests/utils/VectorMaker.cpp (hunk 2 above)
Comment (leoluan2009) [reply velox/vector/tests/utils/VectorMaker.cpp]: Done, Thanks

**Comment by leoluan2009** [comment]
Hi, @majetideepak @kgpai, I fix all comments, would you please review again? Thanks!

**Pull request closed without merging**